### PR TITLE
Add conversion between qcodes legacy DataSet and xarray format

### PR DESCRIPTION
**Pull request by peendebak**
Changes proposed in this pull request:

- Add conversion methods between the QCoDeS legacy dataset and xarray Dataset

A minimal example:
```
from qcodes.tests.legacy.data_mocks import DataSet1D, DataSet2D
qd = DataSet1D()
xds = qd.to_xarray()  
print(xds)
qds = DataSet.from_xarray(xds)
```

@jenshnielsen The goal of this PR is to allow conversion between the easy cases (e.g. 1D, 2D). Since the two datasets have slightly different formats, we do not cover all cases. Can you comment on whether this functionality can be added to qcodes and if the structure of the PR is good?

### CODE REVIEW

**Comment by jenshnielsen** [comment]
@peendebak Thanks, I will have a look. Happy to merge this to QCoDeS. I have edited the description to clarify that this is about the legacy `data.dataset` and not the current QCoDeS dataset.

File: qcodes/data/data_array.py [hunk 1]
```diff
@@ -529,3 +532,50 @@ def fraction_complete(self):
...
+    else:
+        if data_array.set_arrays:
+            data_dictionary["dims"] = tuple([a.array_id for a in data_array.set_arrays])
+            data_dictionary["depends_on"] = data_dictionary["dims"]
+        data_dictionary["data"] = data_array.ndarray
+
+    return data_dictionary
+
+
+if __name__ == '__main__':
```
Comment (astafan8) [inline qcodes/data/data_array.py:578]: let's remove this if block, it's contents are in the tests (as far as i understand), and qcodes submodules are not really meant to be executable))

**Review by astafan8 — APPROVED** [review]
@peendebak thanks for the contribution (and the typing, and the tests)! if this helps for the users of the legacy DataSet, let's get it in :)

File: qcodes/data/data_array.py (hunk 1 above)
Comment (peendebak) [reply qcodes/data/data_array.py]: Done!

**Comment by peendebak** [comment]
@astafan8 I addressed the comments, and added some more tests. The build seems to be failing due to a general error.

**Comment by astafan8** [comment]
> The build seems to be failing due to a general error.

@peendebak  what do you mean with "general error"? the error here https://github.com/QCoDeS/Qcodes/pull/2938/checks?check_run_id=2399670141 seems to be a typo in the assert method call:
```
>       self.asserEqual(qd.default_parameter_array().shape, xarray_dataset.tmin.shape)
E       AttributeError: 'TestDataSet' object has no attribute 'asserEqual'
```
it should be assertEqual, with "t" :)

**Pull request merged**